### ISSUE TITLE
how to load YUEFU pretrain model weights

### ISSUE THREAD

**Issue by luoda888** [bojone/bert4keras#224]
提问时请尽可能提供如下信息：

### 基本信息
- 你使用的**操作系统**:  ubuntu 18.04
- 你使用的**Python**版本:  Python 3.7
- 你使用的**Tensorflow**版本:  TF 1.14.0
- 你使用的**Keras**版本:  2.4.3
- 你使用的**bert4keras**版本:  0.8.8
- 你使用纯**keras**还是**tf.keras**:  tf.keras
- 你加载的**预训练模型**: YUEFU

### 核心代码
```python
# 请在此处贴上你的核心代码。
# 请尽量只保留关键部分，不要无脑贴全部代码。
```

### 输出信息
```shell
# 请在此处贴上你的调试输出
```

### 自我尝试
不管什么问题，请先尝试自行解决，“万般努力”之下仍然无法解决再来提问。此处请贴上你的努力过程。

How to build config.json for YUEFU model? just like ELE?~ Thanks
and how to init your open-scored weights wonezha, show that position-embedding not found

**Comment by bojone** [comment]
1. yuefu我还没关注到，有空看看；
2. `build_transformer_model`时传入`model=nezha`就可以加载NEZHA类模型。

**Comment by luoda888** [comment]
另反馈一些问题。
1. DataLoader在tf2会很慢，改成tf 1.14.0就速度正常了 tf2需要使用dataset策略或者手动构造数据集才能保证加载速度
2.
在Transformer使用convert_torch_to_tf转换后，albert文件无法载入save_weights函数，看能否下一个版本让bert4keras自适应hugging-face的torch权重，内置一个转换工具，这样可以支持更丰富的多语言模型，从huggingface引流，构建torch2keras的生态
3.
tf2在调用tensorflow_addons时，需要维护keras=2.4.3的版本，否则会出现由于tf.keras和keras混编导致的错误，这一点在全局指定keras
& tf_keras时也无效
4. 建议加入生成式文本阅读的benchmark, 已经把抽取式的mode写在那了干脆加上来嘛
5. 考虑加入bert的mixup么，线下测试了效果不错

苏剑林(Jianlin Su) <notifications@github.com> 于2020年10月12日周一 下午4:35写道：

**Comment by bojone** [comment]
谢谢你的建议。

1. DataLoader是哪里？bert4keras没有DataLoader呀。
2.torch转tf没有通用的方法，只能一个个实现，然后一个个转，这个没办法～
3.设置环境变量TF_KERAS=1也不行吗？
4.不明白你的意思？加到哪里？
5.文本的mixup的标准做法是怎样呢？有没有比较standard的论文可以参考下？

**Comment by luoda888** [comment]
1. 在大多数例子中，数据集input使用的都是fit-fit_generator的方法，这种在tf2里会变得非常缓慢（tf2.2
keras2.3.1 & 2.4.3）
    我们可以使用直接加载并在fit函数中控制batch_size的方法替代fit_generator的缓慢，实测效果非常稳定，且能够适应全版本
example:
def gen_data(df, padding=0):
    x1, x2 = [],[]
    for i in df:
        a, b = tokenizer.encode(i, maxlen=maxlen)
        x1.append(np.array(a))
        x2.append(np.array(b))

    x1 = pad_sequences(x1, maxlen=maxlen, value=int(padding),
padding='post')
    x2 = pad_sequences(x2, maxlen=maxlen, value=int(padding),
padding='post')
    return [x1, x2]

2. 可以考虑提供Hugging-face的下载器，如wget后放置在同一个文件夹，再利用如下脚本进行转换
这个是考虑到很多时候大文件没办法很好的download下来，有一些入门者对于下载权重及使用会比较困难
from transformers import pipeline
import transformers as tfs
import numpy as np
import tensorflow.compat.v1 as tf
import torch
from tqdm import tqdm

from transformers import BertModel, AutoModel, AlbertModel, ElectraModel,
RobertaModel, GPT2Model, DistilBertModel

def convert_pytorch_checkpoint_to_tf(model: BertModel, ckpt_dir: str,
model_name: str):

    """
    :param model:BertModel Pytorch model instance to be converted
    :param ckpt_dir: Tensorflow model directory
    :param model_name: model name
    :return:
    Currently supported HF models:
        Y BertModel
        N BertForMaskedLM
        N BertForPreTraining
        N BertForMultipleChoice
        N BertForNextSentencePrediction
        N BertForSequenceClassification
        N BertForQuestionAnswering
    """

    tensors_to_transpose = ("dense.weight", "attention.self.query",
"attention.self.key", "attention.self.value")

    var_map = (
        ("layer.", "layer_"),
        ("word_embeddings.weight", "word_embeddings"),
        ("position_embeddings.weight", "position_embeddings"),
        ("token_type_embeddings.weight", "token_type_embeddings"),
        (".", "/"),
        ("LayerNorm/weight", "LayerNorm/gamma"),
        ("LayerNorm/bias", "LayerNorm/beta"),
        ("weight", "kernel"),
    )

    if not os.path.isdir(ckpt_dir):
        os.makedirs(ckpt_dir)

    state_dict = model.state_dict()

    def to_tf_var_name(name: str):
        for patt, repl in iter(var_map):
            name = name.replace(patt, repl)
        return "bert/{}".format(name)

    def create_tf_var(tensor: np.ndarray, name: str, session: tf.Session):
        tf_dtype = tf.dtypes.as_dtype(tensor.dtype)
        tf_var = tf.get_variable(dtype=tf_dtype, shape=tensor.shape,
name=name, initializer=tf.zeros_initializer())
        session.run(tf.variables_initializer([tf_var]))
        session.run(tf_var)
        return tf_var

    tf.reset_default_graph()
    with tf.Session() as session:
        for var_name in tqdm(state_dict):
            tf_name = to_tf_var_name(var_name)
            torch_tensor = state_dict[var_name].numpy()
            if any([x in var_name for x in tensors_to_transpose]):
                torch_tensor = torch_tensor.T
            tf_var = create_tf_var(tensor=torch_tensor, name=tf_name,
session=session)
            tf.keras.backend.set_value(tf_var, torch_tensor)
            tf_weight = session.run(tf_var)
#             print("Successfully created {}: {}".format(tf_name,
np.allclose(tf_weight, torch_tensor)))

        saver = tf.train.Saver(tf.trainable_variables())
        saver.save(session, os.path.join(ckpt_dir, model_name.replace("-",
"_") + ".ckpt"))

for i in model_list:
    path = "../chinese-tfs-weights/{}".format(i)
    print(path)
    if not os.path.exists("{}/pytorch_model.bin".format(path)):
        continue
    try:
        model = AutoModel.from_pretrained(
                pretrained_model_name_or_path=path,
                state_dict=torch.load("{}/pytorch_model.bin".format(path)),
        )
    except:
        print("ERROR:", i)
        model = BertModel.from_pretrained(
                pretrained_model_name_or_path=path,
                state_dict=torch.load("{}/pytorch_model.bin".format(path)),
        )

    convert_pytorch_checkpoint_to_tf(model=model, ckpt_dir=path,
                                     model_name='tf_model')

但是如hfl系列roberta模型，在bert4keras中加载不能简单传入model='roberta',应传入model='bert'

3.
设置KERAS和TF_KERAS都无效，今日DEBUG发现，可能在bert4keras的文档中需要提及不能使用系统默认的KERAS模块，必须从bert4keras中调用
keras package,很大一部分不匹配的错误都是从这来的

4. 加到example/task_reading_comprehension_by_seq2seq.py
<https://github.com/bojone/bert4keras/blob/master/examples/task_reading_comprehension_by_seq2seq.py>
里，
代码中设置了self.mode == 'extractive': 但是并没有考虑生成式的
5. 参考textaugment库中 mixup的做法，以及paipaidai的做法，paper：
https://github.com/dsfsi/textaugment
https://openreview.net/forum?id=r1Ddp1-Rb
https://arxiv.org/abs/1905.08941

苏剑林(Jianlin Su) <notifications@github.com> 于2020年10月13日周二 上午11:43写道：

**Comment by bojone** [comment]
fit_generator是过时了，有空改过来才行。

设置TF_KERAS=1要放到导入bert4keras之前。
